### ISSUE TITLE
Underlying Xgboost Exception in TPOTClassifier.Fit 

### ISSUE THREAD

**Issue by gvyshnya** [rhiever/tpot#545]
TPOTClassifier.Fit method inclines to select _xgboost_ as an optimal model algorithm for the MNIST data set (vs. other locally installed modellers) but then it fires the error below

> _"[23:44:15] c:\dev\libs\xgboost\dmlc-core\include\dmlc\./logging.h:235: [23:44:15] C:\dev\libs\xgboost\src\tree\updater_colmaker.cc:162: Check failed: (n) > (0U) colsample_bytree=1 is too small that no feature can be included"_

In the evidence of the problem, I have two questions
- Is there a way to conditionally exclude xgboost (or equally any other algorithm) from the grid search that TPOTClassifier does?
- What is the best workaround for the problem above?

## Process to reproduce the issue

The issue is reproduced if you run the code below in your Python 3.6-based environment:

`import pandas as pd
import datetime as dt
from sklearn.cross_validation import train_test_split
import tpot

start_time = dt.datetime.now()

print("Started at ", start_time)
print("Reading data...")
train_data = pd.read_csv('data/train.csv')
test_data = pd.read_csv('data/test.csv')

X = train_data.drop('label', axis=1).values
y = train_data.loc[:, 'label'].values

X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    train_size=0.75,
                                                    test_size=0.25)

my_tpot = tpot.tpot.TPOTClassifier(generations=10)
print("Completed initializing the auto-ml classifier at ...", dt.datetime.now())

my_tpot.fit(X_train, y_train)
print("Completed fitting the model at ...", dt.datetime.now())
print(my_tpot.score(X_test, y_test))

my_tpot.export('exported_pipeline.py')

end_time = dt.datetime.now()
elapsed_time = end_time - start_time
print("Finished AutoML to generate the best model and pipeline")
print("Elapsed time: ", elapsed_time)`

**Note:** the training and testing datasets relate to the famous MNIST data that can be downloaded at https://www.kaggle.com/c/digit-recognizer/data, for instance

## Expected result

TPOTClassifier.Fit works smoothly, and it eventually generates the best model for the MNIST classification problem. That model will be also able to export as a Python module.

## Current result

The error message is fired per the listing below

> _"[23:44:15] c:\dev\libs\xgboost\dmlc-core\include\dmlc\./logging.h:235: [23:44:15] C:\dev\libs\xgboost\src\tree\updater_colmaker.cc:162: Check failed: (n) > (0U) colsample_bytree=1 is too small that no feature can be included"_

The script then stops, and no optimal model is fitted and exported as a Python module.

## Notes
I tried the above-mentioned code before I installed xgboost into Anaconda on my Window 7 64-bit machine (Python 3.6 runtime used by Anaconda). At that time, it did not produce any error, and the best model to be suggested was Logistic Regression model.

**Comment by weixuanfu** [comment]
The answer to the first question is using the customized configuration of operators and their parameters.
 Please check the [link](http://rhiever.github.io/tpot/using/#customizing-tpots-operators-and-parameters) for more details.

The second question of this issue related to the issue #449 in tpot repo and [the issue 2349](https://github.com/dmlc/xgboost/issues/2349) in xgboost repo. In general, the logging cannot be disabled  with `silent=True` in xgboost API.

**Comment by gvyshnya** [comment]
@weixuanfu : thanks for detailed and prompt explanation. 

Just wanted to clarify the practical steps on my end.
1. Since no model is generated by TPOT on my premise due to the xgboost-driven issue referred, shall I simply exclude in the customized configuration?
2. Otherwise, shall I try to tweak xgboost settings in the customized configuration to find some setup where it does not crash and allows for a model to be generated by TPOT eventually?

Thank you in advance.

**Comment by weixuanfu** [comment]
Yes, you may need exclude xgboost from [built-in configuration](https://github.com/rhiever/tpot/blob/master/tpot/config/classifier.py) to avoid the error. Or you could try to run the codes with default configuration in Linux or macOS where the error message would still be printed out but TPOT optimization process would not be stopped.

**Comment by shirishr** [comment]
Got same error message ( colsample_bytree=1 is too small that no feature can be included)
But I was running TPOTClassifier(generations=10, population_size=50, verbosity=2, n_jobs=4) and then noticed that xgboost configuration calls out
    'nthread': [1]
Is this relevant?
Am continuing testing by excluding the Xgboost. Will keep you posted.

**Comment by rhiever** [comment]
>'nthread': [1]

This particular line is telling XGBoost not to use its built-in multithreading because it causes issues with TPOT. Until the next major XGBoost release, the default behavior of XGBoost was to use multithreading.

**Comment by rhiever** [comment]
Closing this issue. Please feel free to re-open or file a new issue if you have any further questions or comments.
